add introduction text, rename to Referenced Token (#20)
vcstuff/draft-looker-jwt-cwt-status-list: draft-looker-jwt-cwt-status-list.md

# Introduction

> JSON Web Tokens (JWTs) {{RFC7519}} and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered suspended or revoked by its issuer.

Reviewer: ```suggestion
JSON Web Tokens (JWTs) {{RFC7519}} and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
```

to match with the rest of terminology

> This document defines a status list in JWT and CWT representation that describes the individual statuses of multiple subject tokens. The statuses of each subject token are conveyed via a bit array in the status list. Each subject token during issuance is allocated an index which represents a position within this bit array and the value of the bit(s) at this position correspond to the subject tokens status. The document also defines how an issuer of a subject token in JWT or CTW representation references a status list. Status lists may be composed for expressing a range of status types, the document defines two configurations for the most common use cases as well as an extensibility mechanism for custom status list types. The status list may be used by an issuer in the Issuer-Holder-Verifier model, as described in (XXX) to express the status of verifiable credentials (Subject Tokens) issued by an issuer.

Reviewer: ```suggestion
This document defines a status list in JWT and CWT representations that describes the individual statuses of multiple subject tokens. The statuses of each subject token are conveyed via a bit array in the status list. Each subject token during issuance is allocated an index which represents a position within this bit array and the value of the bit(s) at this position correspond to the subject tokens status. The document also defines how an issuer of a subject token in JWT or CTW representation references a status list. Status lists may be composed for expressing a range of status types, the document defines two configurations for the most common use cases as well as an extensibility mechanism for custom status list types. The status list may be used by an issuer in the Issuer-Holder-Verifier model, as described in (XXX) to express the status of verifiable credentials (Subject Tokens) issued by an issuer.
```

Reviewer 2: ```suggestion
This document defines a status list in JWT and CWT representation that describes the individual statuses of multiple subject tokens. The statuses of each subject token are conveyed via a bit array in the status list. Each subject token during issuance is allocated an index which represents a position within this bit array and the value of the bit(s) at this position correspond to the subject tokens status. The document also defines how an issuer of a subject token in JWT or CWT representation references a status list. Status lists may be composed for expressing a range of status types, the document defines two configurations for the most common use cases as well as an extensibility mechanism for custom status list types. The status list may be used by an issuer in the Issuer-Holder-Verifier model, as described in (XXX) to express the status of verifiable credentials (Subject Tokens) issued by an issuer.
```

> This document defines a Status List in JWT and CWT representation that describes the individual statuses of multiple Subject Tokens. The Status List contains binary data, where each Subject Token corresponds with a certain position within a bit array(also known as bit vector, string or list). The document also defines how an issuer of a Subject Token in JWT or CTW representation references a Status List to convey status data like suspension or revocation of the Subject Token and related information. Status Lists may be composed for a range of various Status Types with one or more bits, the document defines two configurations for the most common use cases as well as an extensibility mechanism for custom Status List Types. The Status List may be used by a Status List Provider in the Issuer-Holder-Verifier model, as described in {{RFC-XXXX}} to express the status of verifiable credentials (Subject Tokens) issued by an issuer.

Reviewer 2: ```suggestion
This document defines a status list in JWT and CWT representation that describes the individual statuses of multiple subject tokens. The statuses of each subject token are conveyed via a bit array in the status list. Each subject token during issuance is allocated an index which represents a position within this bit array and the value of the bit(s) at this position correspond to the subject tokens status. The document also defines how an issuer of a subject token in JWT or CTW representation references a status list. Status lists may be composed for expressing a range of status types, the document defines two configurations for the most common use cases as well as an extensibility mechanism for custom status list types. The status list may be used by an issuer in the Issuer-Holder-Verifier model, as described in (XXX) to express the status of verifiable credentials (Subject Tokens) issued by an issuer.
```

Reviewer 2: Bit of a suggested revision to the text here.

Author: @tplooker Looking at other RFCs like JWT, I see that things that are mentioned in the terminology are capitalized throughout the text. Why does this not apply here?

> This document defines a Status List in JWT and CWT representations that describes the individual statuses of multiple Referenced Tokens. The statuses all Referenced Tokens are conveyed via a bit array in the Status List. Each Referenced Token during issuance is allocated an index which represents a position within this bit array and the value of the bit(s) at this position correspond to the Referenced Token's status. The document also defines how an issuer of a Referenced Token in JWT or CWT representation references a Status List Token. Status Lists may be composed for expressing a range of Status Types, the document defines basic Status Types for the most common use cases as well as an extensibility mechanism for custom Status Types. The Status List Token may be used by an issuer in the Issuer-Holder-Verifier model, as described in (XXX) to express the status of verifiable credentials (Referenced Tokens) issued by an issuer.

Reviewer 3: ```suggestion
This document defines a Status List in JWT and CWT representations that describes the individual statuses of multiple Referenced Tokens. The statuses of all Referenced Tokens are conveyed via a bit array in the Status List. Each Referenced Token is allocated an index during issuance which represents a position within this bit array and the value of the bit(s) at this position correspond to the Referenced Token's status. The document also defines how an issuer of a Referenced Token in JWT or CWT representation references a Status List Token. Status Lists may be composed for expressing a range of Status Types, the document defines basic Status Types for the most common use cases as well as an extensibility mechanism for custom Status Types. The Status List Token may be used by an issuer in the Issuer-Holder-Verifier model, as described in (XXX) to express the status of verifiable credentials (Referenced Tokens) issued by an issuer.
```

The following diagram depicts the basic conceptual relationship.

~~~ ascii-art

+---------------+                     +---------------+
|               |      References     |               |
|               |-------------------->|               |
| Subject Token |                     |  Status List  |
|(JWT/CWT Based)|                     |(JWT/CWT Based)|
|               |  Describes Status   |               |
|               |<--------------------|               |
+---------------+                     +---------------+

~~~

## Rationale

Revocation mechanisms are an essential part for most identity ecoosystems. In the past, revocation of X.509 TLS certificates has been proven difficult as traditional certificate revocation lists (CRLs) have limited scalability and the Online Certificate Status Protocol (OCSP) has additional privacy risks as the client is leaking the requested website to a third party. OSCP stapling is adressing some of these problems at the cost of less up-to-date data. Modern approaches use accumulator-based revocation registries and Zero-Knowledge-Proofs to accomodate for this privacy gap but face scalability issues again.

> The approach of this specification seeks to find a balance between scalability, security and privacy by minimizing the status information to mere bits and compressing the resulting binary data. Thereby a Status List may contain statuses of 100.000 or more Subject Tokens, but still remain relatively small. Placing large amounts of Subject Tokens into the same list also enables a herd privacy towards the Issuer or Status List Provider.

Reviewer 3: Is the point that 100000 or more subject tokens makes it reasonable privacy preserving?

# Conventions and Definitions

{::boilerplate bcp14-tagged}

# Terminology

Status List
 A token in JWT or CWT representation that expresses the statuses of many Subject Tokens.

Subject Token
 A token in JWT or CWT representation which contains a reference to a Status List. The information from the Stataus List may give a verifier additional information about up-to-date status of the Subject Token.

Status List Provider

>  An entity that hosts the Status List as a resource for potential verifiers. The Status List Provider may be the issuer of the Status List but may also be outsourced to a trusted third party.

Reviewer 2: Not super convinced of this, I think this is pending resolution around whether we agree if the status list provider != issuer otherwise I think issuer suffices?

Author: This relates to #10 and is in line with the issuer-holder-verifier model described in the SD-JWT spec

TODO elaborate that a status list only gives information about the maximum number of possible statuses that a list conveys, issuers are recommended to pre-allocate lists, use dead entries that are never assigned or other obfuscation mechanisms

## Malicious Issuers

> TODO elaborate on issuers generating unique status lists per Referenced Token that do not offer herd privacy

Reviewer 3: Can we keep TODOs as issues, pls.? It makes it easier to grasp the TODOs for a certain draft.

Author: these TODOs are partly resolved in other PRs and I would like to not change them in this PR as this resulsts to even more merge conflicts
